### PR TITLE
Refactor to remove mbl-scratch dependency.

### PR DESCRIPTION
**Pull request by petedyerarm**
Added files to enable the use of pytest to build test images and run
tests on either LXC or device under test.

Added conftest to output LAVA compatible strings instead of scraping the
XML output in seperate scripts.

Added fixtures into conftest to include helpers common to every test,

Added pytest files to build test applications instead of using the
pre-built images from mbl-scratch.

Changed the core component tests into a pytest that is run inside the
LXC container, thus removing the need to inject the tests directly into
the rootfs on the device under test.

Converted systemd test from bash to pytest. Removed mnt.flags test as 
that partition is no longer used.

JIRA IOTMBL-854.

### CODE REVIEW

**Comment by petedyerarm** [comment]
SYNC with https://github.com/ARMmbed/mbl-tools/pull/291/

File: ci/lava/conftest.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    """Fixture for --dut."""
+    return request.config.getoption("--dut")
+
+
+class Execute_Helper:
+    """Class to provide a wrapper for executing commands as a subprocess."""
+
+    @staticmethod
+    def _print(data):
+        # print(data)
```
[the remark this answers is not in the document]
Comment (jjohnson-arm) [reply ci/lava/conftest.py]: Or controlled by global?

File: ci/lava/conftest.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        pass
+
+    @staticmethod
+    def _execute_command(command):
+        """Execute the provided command list.
+
+        Executes the command and returns the erro code, stdout and stderr.
```
Comment (jjohnson-arm) [inline ci/lava/conftest.py]:
```suggestion
        Executes the command and returns the error code, stdout and stderr.
```

File: ci/lava/conftest.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+@pytest.fixture
+def execute_helper():
+    """Fixture to return the helper class."""
+    return Execute_Helper
+
+
+@pytest.fixture
+def dut_addr(dut, execute_helper):
+    """Fixture to calculate the DUT IP address."""
```
Comment (jjohnson-arm) [inline ci/lava/conftest.py]: Add comment explaining dut parameter - "auto" and regexp options?

File: ci/lava/tests/test-build-test-images.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
+#!/usr/bin/env python3
```
Comment (jjohnson-arm) [inline ci/lava/tests/test-build-test-images.py]: Again - maybe better to refer to as tutorial images? Or at least application images or something. Test images seems wrong, unless you think we are going to increase scope to boot loaders etc.

File: ci/lava/tests/test-core-component.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+class Test_Core_Component_DUT:
+    """Class to encapsulate the testing of core components on a DUT."""
+
+    def test_setup(self, dut_addr, execute_helper):
+        """Copy the test specific parts, generating items as required."""
+        execute_helper._send_mbl_cli_command(
+            [
+                "put",
+                "/tmp/user-sample-app-package_1.0_any.ipk",
+                "/scratch/app-lifecycle-manager-test-package_1.0_any.ipk",
```
Comment (jjohnson-arm) [inline ci/lava/tests/test-core-component.py]: Again - this should either have class strings (or globals) that define /tmp and /scratch - or have defaults and paramters to control them.

File: ci/lava/conftest.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+        Execute_Helper._print("error:")
+        Execute_Helper._print(error)
+        Execute_Helper._print("returnCode:")
+        Execute_Helper._print(p.returncode)
+        Execute_Helper._print("execute_command done.")
+
+        return p.returncode, output, error
+
+    @staticmethod
+    def _send_mbl_cli_command(command, addr):
```
Comment (rwalton-arm) [inline ci/lava/conftest.py]: I'm not sure what the public interface of this class is. looking at the test examples it looks like this `_send_mbl_cli_command` method. If that is the case we should remove the leading underscore, as having it there implies this method is an implementation detail.

File: ci/lava/conftest.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    return request.config.getoption("--host_tutorials_dir")
+
+
+@pytest.fixture
+def venv(request):
+    """Fixture for --venv."""
+    return request.config.getoption("--venv")
+
+
+class Execute_Helper:
```
Comment (rwalton-arm) [inline ci/lava/conftest.py]: Convention is to not use underscores in class names. Should be `ExecuteHelper`, same comment for other classes.

File: ci/lava/tests/test-build-tutorial-images.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+USER_SAMPLE_TAR = "user-sample-app-package_1.0_any.ipk.tar"
+MULTI_APP_ALL_GOOD_TAR = "mbl-multi-apps-update-package-all-good.tar"
+MULTI_APP_ONE_FAIL_RUN_TAR = "mbl-multi-apps-update-package-one-fail-run.tar"
+MULTI_APP_ONE_FAIL_INSTALL_TAR = (
+    "mbl-multi-apps-update-package-one-fail-install.tar"
+)
+
+
+class Test_Build_Test_Images:
+    """Class to encalpulate building images for testing."""
```
Comment (rwalton-arm) [inline ci/lava/tests/test-build-tutorial-images.py]:
```suggestion
    """Class to encapsulate building images for testing."""
```

File: ci/lava/conftest.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+    @staticmethod
+    def _send_mbl_cli_command(command, addr):
+
+        err = 1
+        output = ""
+        error = ""
+
+        if addr != "":
+            cliCommand = ["mbl-cli", "--address", addr]
```
Comment (rwalton-arm) [inline ci/lava/conftest.py]: Convention is to use all lowercase with underscores in variable names, i.e `cli_command`

File: ci/lava/conftest.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+    """Fixture to return the helper class."""
+    return Execute_Helper
+
+
+@pytest.fixture
+def dut_addr(dut, execute_helper):
+    """Fixture to calculate the DUT IP address."""
+    dut_addr = ""
+
+    returnCode, list_output, error = execute_helper._execute_command(
```
Comment (rwalton-arm) [inline ci/lava/conftest.py]:
```suggestion
    return_code, list_output, error = execute_helper._execute_command(
```

File: ci/lava/dependencies/create-python-environment.yaml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+run:
+    steps:
+        # Create and activate the python venv
+        - python3 -m venv $virtual_env
+        - . $virtual_env/bin/activate
+
+        # Upgrade version of pip3
+        - pip3 install --upgrade pip
+
+        # Download pytest. Do this in two stages to the files are available for installing on the DUT as well
```
[the remark this answers is not in the document]
Comment (petedyerarm) [reply ci/lava/dependencies/create-python-environment.yaml]: changed.

File: ci/lava/dependencies/create-python-environment.yaml [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+        #
+        - pip3 download -d /tmp/pytest pytest
```
[the remark this answers is not in the document]
Comment (petedyerarm) [reply ci/lava/dependencies/create-python-environment.yaml]: the lxc container is re-created for every test job, so this is OK.

File: ci/lava/dependencies/test-install-pytest-to-target.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""
+Script to install test dependencies onto the DUT.
+
+The test is run using pytest and communicates with the DUT using the
+mbl-cli which is previously installed as part of the wider LAVA test.
+
+This script also assumes they pytest has been downloaded
+into /tmp, also as part of the wider LAVA test. As the LAVA test is using
```
[the remark this answers is not in the document]
Comment (petedyerarm) [reply ci/lava/dependencies/test-install-pytest-to-target.py]: Changed the hard coded paths to be parameters that can be specified to pytest via the job pipelines.

File: ci/lava/dependencies/test-install-pytest-to-target.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+            match = pattern.match(line)
+            if match:
+                copy_success = True
+                break
+        assert copy_success
+
+    def test_create_python_virtual_environment_on_target(
+        self, execute_helper, dut_addr
+    ):
+        """Create the python virtual environment on the DUT using mbl-cli."""
```
[the remark this answers is not in the document]
Comment (petedyerarm) [reply ci/lava/dependencies/test-install-pytest-to-target.py]: changed

File: ci/lava/dependencies/create-python-environment.yaml [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+metadata:
+    format: Lava-Test Test Definition 1.0
+    name: Create Python environment including pytest into the LXC container.
+    description: Python Environment Creation.
+
+parameters:
+    # virtual_env specifies the name of the virtual environment
+    #
+    virtual_env:
```
[the remark this answers is not in the document]
Comment (petedyerarm) [reply ci/lava/dependencies/create-python-environment.yaml]: Changed the hard coded paths to be parameters that can be specified to pytest via the job pipelines.

File: ci/lava/conftest.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (petedyerarm) [reply ci/lava/conftest.py]: Added comment about usage.

File: ci/lava/conftest.py (hunk 2 above)
Comment (petedyerarm) [reply ci/lava/conftest.py]: changed

File: ci/lava/conftest.py (hunk 3 above)
Comment (petedyerarm) [reply ci/lava/conftest.py]: Added comment.

File: ci/lava/tests/test-build-test-images.py (hunk 4 above)
Comment (petedyerarm) [reply ci/lava/tests/test-build-test-images.py]: changed.

File: ci/lava/tests/test-core-component.py (hunk 5 above)
Comment (petedyerarm) [reply ci/lava/tests/test-core-component.py]: Changed the hard coded paths to be parameters that can be specified to pytest via the job pipelines.

File: ci/lava/conftest.py (hunk 6 above)
Comment (petedyerarm) [reply ci/lava/conftest.py]: changed.

File: ci/lava/conftest.py (hunk 7 above)
Comment (petedyerarm) [reply ci/lava/conftest.py]: changed

File: ci/lava/tests/test-build-tutorial-images.py (hunk 8 above)
Comment (petedyerarm) [reply ci/lava/tests/test-build-tutorial-images.py]: changed

File: ci/lava/conftest.py (hunk 9 above)
Comment (petedyerarm) [reply ci/lava/conftest.py]: changed

File: ci/lava/conftest.py (hunk 10 above)
Comment (petedyerarm) [reply ci/lava/conftest.py]: changed

**Pull request merged**